### PR TITLE
Add a `modular` argument to `executable`

### PR DESCRIPTION
**Pull request by lamefun**
This complements the `shared_module` feature. I know that how I add it to compiler is all wrong, so I need advice.

### CODE REVIEW

File: mesonbuild/build.py [hunk 1]
```diff
@@ -1160,12 +1161,21 @@ def __init__(self, name, subdir, subproject, is_cross, sources, objects, environ
         # The import library that GCC would generate (and prefer)
         self.gcc_import_filename = None
 
+        # check for modularity
+        self.is_modular = False
+        if 'modular' in kwargs and kwargs['modular']:
+            self.is_modular = True
+        if 'implib' in kwargs and kwargs['implib']:
+            self.is_modular = True
+        if self.is_modular and ('implib' in kwargs and kwargs['implib'] == False):
```
Comment (ignatenkobrain) [inline mesonbuild/build.py]: __[Flake8]__

comparison to False should be 'if cond is False:' or 'if not cond:'

File: mesonbuild/build.py [hunk 2]
```diff
@@ -1160,12 +1161,21 @@ def __init__(self, name, subdir, subproject, is_cross, sources, objects, environ
...
+            raise InvalidArguments('"implib" must not be set to false for modular executables')
+
         # if implib appears, this target is linkwith:-able, but that only means
         # something on Windows platforms.
         self.is_linkwithable = False
-        if 'implib' in kwargs and kwargs['implib']:
+        if self.is_modular:
             implib_basename = self.name + '.exe'
-            if not isinstance(kwargs['implib'], bool):
+            if kwargs.get('implib') != None and not isinstance(kwargs['implib'], bool):
```
Comment (ignatenkobrain) [inline mesonbuild/build.py]: __[Flake8]__

comparison to None should be 'if cond is not None:'

File: docs/markdown/Reference-manual.md [hunk 3]
```diff
@@ -999,6 +1002,9 @@ This is useful for building modules that will be `dlopen()`ed and
 hence may contain undefined symbols that will be provided by the
 library that is loading it.
 
+If you want an [`executable`](#executable) to be able to load shared
+modules, you will need to set its `modular` argument to `true`.
+
```
Comment (jon-turney) [inline docs/markdown/Reference-manual.md]: This is not a good description of what this argument does, and `man ld` manages better.

`-export-dynamic` is _only_ needed if you want to load a shared module _which references symbols in the executable_.

Given that, the name `modular` for this argument is not well-chosen.

See test case common/156 (which uses the compiler option `-rdynamic `which causes this linker option to be used), which should probably be updated to use this argument. (and compare with test case common/125, which does not use or need this option)

**Comment by lamefun** [comment]
I realized that I want separate branches for stuff, but GitHub doesn't allow me to change the branch I'm sending the patch from, so I guess I'll have to open a new one.

**Pull request closed without merging**